### PR TITLE
Attach source code locations to Irmin's log lines

### PR DESCRIPTION
**Pull request by CraigFe**
As discussed offline, this PR converts Irmin's use of the `Logs` API to go via a small internal PPX (`ppx_irmin.internal`). This PPX attaches source locations to the log entries using Logs' [tags API](https://erratique.ch/software/logs/doc/Logs/Tag/index.html). The tag specification is exported within the new `Irmin.Export_for_backends.Logging` module, which also provides a default `Logs.reporter` that is sensitive to these tags. If an entry doesn't have an attached source code position (i.e. the log line is from another library), the "source" name is printed instead (as before):

image

In terms of _using_ the newly-written reporter, this PR takes the initial step of replacing `Irmin_test.reporter` with it (which covers most of our test suite), but it doesn't try to unify all of the reporters in the codebase yet. It seems better to do that as part of a separate refactoring pass, as there are some fiddly edge-cases involved.

We should also provide some sort of dynamic white-listing / black-listing mechanism for log entries – I haven't done that here.

### CODE REVIEW

File: src/irmin-git/atomic_write.ml [hunk 1]
```diff
@@ -233,9 +232,10 @@ module Make (K : Key) (G : Git.S) = struct
     | _ -> false
 
   let test_and_set t r ~test ~set =
-    Log.debug (fun f ->
-        let pp = Fmt.option ~none:(Fmt.any "<none>") (Irmin.Type.pp Val.t) in
-        f "test_and_set %a: %a => %a" pp_branch r pp test pp set);
+    [%log.debug fun f ->
+      let pp = Fmt.option ~none:(Fmt.any "<none>") (Irmin.Type.pp Val.t) in
+      f "test_and_set %a: %a => %a" pp_branch r pp test pp set]
+    ;
```
Comment (CraigFe) [inline src/irmin-git/atomic_write.ml:238]: This is one of two cases where the new syntax is longer than the old one, and both are due to the same formatting weirdness for code of the form `[%ext fun ... -> ...]; ...`.

I've opened an OCamlformat issue about this [here](https://github.com/ocaml-ppx/ocamlformat/issues/1819).

**Review by Ngoguey42 — APPROVED** [review]

**Comment by CraigFe** [comment]
Thanks @Ngoguey42 for the review. Rebased. I'll merge once the CI is happy.
